### ISSUE TITLE
How to integrate scala_import

### ISSUE THREAD

**Issue by gkossakowski** [bazelbuild/rules_scala#370]
I'm opening this ticket to discuss details of using `scala_import` with IntelliJ. Having the discussion in public will make it easier to Google for other people.

Scala rules have `scala_import` that is supposed to mirror `java_import`. Another feature of `scala_import` is that it carries information that IntelliJ plugin lookups to determine dependencies.

I'm working on adding support for `scala_import` to bazel-deps and I hit some snags. Questions:
1. Should both java and scala go through `*_import` rules? At the moment, bazel-deps uses `java_library` for exposing java jars (fetched with maven_artifact) and `scala_library` for scala jars. For consistency reasons, it feels like we should just use import rules
2. What's the exact dependency of a `java_import`? is it a file group like `@com_google_guava_guava//jar:file` or `@com_google_guava_guava//jar`? What's better for IntelliJ?
3. How are transitive dependencies supposed to be expressed? My understanding is that they should be exposed as exports for both java and scala dependencies. Is that right?
4. Does resolution of source artifacts work for both java and scala dependencies in IntelliJ?

**Comment by gkossakowski** [comment]
@ittaiz I think you might know answers to these questions. I'm curious how you're using scala_import internally with your own equivalent of bazel-deps.

**Comment by johnynek** [comment]
transitive dependencies can either be `exports` or `runtime_deps` in these cases. bazel-deps has an option for each.

I'm pretty sure intellij won't mind ` java_import` pointing at a `file` because that is exactly what `@foo//jar` is (look at it. it is just a generated BUILD file that has a `java_import(name = "jar", ...`. No magic.

**Comment by gkossakowski** [comment]
yup, for others stumbling into this ticket, a sample generated file:

```
# DO NOT EDIT: automatically generated BUILD.bazel file for maven_jar rule com_google_guava_guava
java_import(
    name = 'jar',
    jars = ['guava-21.0.jar'],
    srcjar = 'guava-21.0-sources.jar',
    visibility = ['//visibility:public']
)

filegroup(
    name = 'file',
    srcs = [
        'guava-21.0.jar',
        'guava-21.0-sources.jar',
    ],
    visibility = ['//visibility:public']
)
```

found in
```
cd $(bazel info output_base )
```

**Comment by ittaiz** [comment]
btw, the strange UX part to me is that you can pass `java_import` multiple jars but only one srcjar.
That is a bug in my eyes. OTH if you allow multiple srcjars how do you correlate?

**Comment by johnynek** [comment]
You disagree with me how?

Bazel-deps has options and I’m open to adding more. As far as I know deps == runtime_deps for a jar. Runtime deps is what we use for compile and runtime scope now. I don’t know maven as well as you. I don’t even know what Pom packaging is. Happy to add more control but we haven’t been burned by using all maven deps as runtime unless they are excluded.

**Comment by ittaiz** [comment]
I don't agree in the sense that I don't think using exports for compile deps is a good idea since it pollutes the graph.
Not sure I understand what you mean by  "deps == runtime_deps for a jar"?

**Comment by johnynek** [comment]
You and I agree about not using exports for compile time dependencies. This is an option for people migrating from sbt or maven that does do that by default. We don’t run with that option. Like you, we have maven dependencies on the runtime path only.

So I think we agree, no?

We may disagree that this should even be an option.

What I meant about deps == runtime_deps for java_import because import is not compiling anything. So, deps are just forwarded to the runtime path, which is exactly the same behavior as runtime deps. Similar for scala_import right?

**Comment by ittaiz** [comment]
:) yes we agree. 
I might, indeed, add a warning to people that this option is not recommended but that's not the heart of it.

no, that's actually not correct. The `deps`, `runtime_deps`, `exports` are used so that we know how to build the correct JavaProvider. 
Concretely:
jars-> participate in compile_jars, transitive_compile_jars, transitive_runtime_jars
exports ->  participate in compile_jars, transitive_compile_jars, transitive_runtime_jars
deps ->  transitive_compile_jars, transitive_runtime_jars
runtime_deps -> transitive_runtime_jars

This distinction is valuable when you want to employ `strict-java-deps` like What Carmi mentioned in the issue I linked above.
It's valuable because the transitive_compile_jars are passed to the compiler and not only the direct compile_jars.

Is that clearer?

**Comment by johnynek** [comment]
Yes. I always forget about the phantom compile time deps (which I still get nervous will increase spurious builds in scala due to how much worse ijar is for scala).
